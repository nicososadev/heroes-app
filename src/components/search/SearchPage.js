import React, { useMemo } from 'react'
import queryString from 'query-string'
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';


export const SearchPage = ({ history }) => {

    const location = useLocation()

    const { q = '' } = queryString.parse( location.search )

    const [ formValue, handleInputChange ] = useForm({ searchField: q })

    const { searchField } = formValue

    const handleSubmit = (e) => {
        e.preventDefault()

        history.push(`?q=${searchField}`)
    }

    const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ])

    return (

        <div>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input 
                            type="text" 
                            placeholder="Find your hero" 
                            className="form-control" 
                            name="searchField"
                            value={searchField}
                            onChange={handleInputChange}
                        />
                        <button type="submit" className="btn btn-outline-primary btn-block mt-1">Search</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />   

                    { 
                    ( q === '')
                    &&
                    <div className="alert alert-info">
                        Search a Hero
                    </div>
                    }

                    { 
                    ( q !== '' && heroesFiltered.length === 0)
                    &&
                    <div className="alert alert-danger">
                        No Hero found
                    </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <div className="animate__animated animate__fadeInUp">
                                <HeroCard
                                    key={hero.id}
                                    {...hero}
                                />
                            </div>
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
