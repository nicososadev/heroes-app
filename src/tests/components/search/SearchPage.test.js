import React from 'react'
import '@testing-library/jest-dom'
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom'
import { SearchPage } from '../../../components/search/SearchPage'

describe('Pruebas en el componente SearchPage', () => {
    
    test('Debe mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search']} >
                <Route path='/search' component={ SearchPage } />
            </MemoryRouter>
        )

        expect( wrapper ).toMatchSnapshot()
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Search a Hero')
    })
    
    test('Debe mostrar a Batman y el input con el valor del queryString', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=batman']} >
                <Route path='/search' component={ SearchPage } />
            </MemoryRouter>
        )

        expect( wrapper.find('input').prop('value') ).toBe('batman')
        expect( wrapper ).toMatchSnapshot()
    })

    test('Debe mostrar un error si no es encuentra el hero', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=no-hero-found']} >
                <Route path='/search' component={ SearchPage } />
            </MemoryRouter>
        )

        expect( wrapper.find('.alert-danger').exists() ).toBe( true )
    })

    test('Debe llamar el push del history', () => {
        
        const history = {
            push: jest.fn()
        }

        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=no-hero-found']} >
                <Route path='/search' component={ () => <SearchPage history={ history } /> } />
            </MemoryRouter>
        )

        wrapper.find('input').simulate(
            'change', 
            { target: {
                name: 'searchField',
                value: 'batman'
            }}
        )

        wrapper.find('form').simulate('submit')

        expect( history.push ).toHaveBeenCalledWith(`?q=batman`)
    })
    
})
