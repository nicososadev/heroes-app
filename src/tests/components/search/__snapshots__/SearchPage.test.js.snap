// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente SearchPage Debe mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchField"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary btn-block mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeInUp"
      >
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small>
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente SearchPage Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchField"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary btn-block mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
